File: ios/Sources/KlinikLabFeature/LabTrendScreen.swift
import Charts
import SwiftUI
import KlinikAPI
import KlinikCore
import KlinikDesign

/// Analyte trends, with the reference band behind them (spec M2).
public struct LabTrendScreen: View {
    @Environment(\.colorScheme) private var scheme

    private let model: LabTrendModel

    @State private var state = LabTrendState()

    public init(model: LabTrendModel) {
        self.model = model
    }

    public var body: some View {
        VStack(spacing: 0) {
            content
        }
        .background(Tokens.Palette.background.resolve(for: scheme))
        .task { await refresh { await model.load() } }
    }

    @ViewBuilder
    private var content: some View {
        switch state.phase {
        case .loading:
            Spacer()
            ProgressView().accessibilityLabel(L10n.string("common.loading"))
            Spacer()

        case .empty:
            MessageState(icon: "chart.xyaxis.line", text: L10n.string("lab.trend.empty"))

        case .notFound:
            MessageState(icon: "questionmark.folder", text: L10n.string("error.notFound"))

        case .failed(let message):
            MessageState(
                icon: Tokens.State.labCritical.iconName,
                text: message,
                retryTitle: L10n.string("common.retry")
            ) {
                await refresh { await model.load() }
            }

        case .loaded:
            loaded
        }
    }

    private var loaded: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                // Above the charts, always. A critical value a doctor has to go
                // looking for is one they can miss.
                if !state.critical.isEmpty {
                    CriticalStrip(results: state.critical)
                }

                Picker("", selection: selection) {
                    ForEach(state.trends) { trend in
                        Text(trend.analyteName).tag(trend.id)
                    }
                }
                .pickerStyle(.menu)
                .accessibilityLabel(L10n.string("lab.trend.analyte"))

                if let trend = state.selectedTrend {
                    TrendChart(trend: trend)
                        .frame(height: 240)

                    TrendSummary(trend: trend)
                }
            }
            .padding(Tokens.Spacing.lg)
        }
    }

    private var selection: Binding<String> {
        Binding(
            get: { state.selected ?? state.trends.first?.id ?? "" },
            set: { id in Task { await refresh { await model.select(id) } } }
        )
    }

    private func refresh(_ work: () async -> Void) async {
        await work()
        state = await model.currentState()
    }
}

struct TrendChart: View {
    @Environment(\.colorScheme) private var scheme

    let trend: AnalyteTrend

    var body: some View {
        Chart {
            // The band goes behind the line, and only when every point was
            // measured against it. A band drawn across points from different
            // ranges would put results on the wrong side of a line they were
            // never compared to.
            if let reference = trend.reference, let low = reference.low, let high = reference.high {
                RectangleMark(
                    yStart: .value(L10n.string("lab.trend.referenceLow"), low),
                    yEnd: .value(L10n.string("lab.trend.referenceHigh"), high)
                )
                .foregroundStyle(Tokens.Palette.successSurface.resolve(for: scheme))
            }

            ForEach(trend.points) { point in
                LineMark(
                    x: .value(L10n.string("measurement.date"), point.measuredAt),
                    y: .value(trend.analyteName, point.value)
                )
                .foregroundStyle(Tokens.Palette.accent.resolve(for: scheme))

                PointMark(
                    x: .value(L10n.string("measurement.date"), point.measuredAt),
                    y: .value(trend.analyteName, point.value)
                )
                // Critical points are red, and the summary underneath says so
                // in words — colour alone carries nothing for a reader who
                // cannot distinguish it (spec section 7).
                .foregroundStyle(colour(for: point.flag).resolve(for: scheme))
            }
        }
        .chartYScale(domain: .automatic(includesZero: false))
        .accessibilityLabel("\(trend.analyteName) (\(trend.unit))")
    }

    private func colour(for flag: LabFlag?) -> ThemedColor {
        switch flag {
        case .critical: return Tokens.Palette.critical
        case .low, .high: return Tokens.Palette.warning
        case .normal, nil: return Tokens.Palette.accent
        }
    }
}

/// The chart in words, for anyone who cannot see it — a plot is invisible to
/// VoiceOver however it is labelled.
struct TrendSummary: View {
    @Environment(\.colorScheme) private var scheme

    let trend: AnalyteTrend

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
            if let latest = trend.points.last {
                HStack(spacing: Tokens.Spacing.sm) {
                    Text("\(L10n.string("measurement.latest")): \(format(latest.value)) \(trend.unit)")
                        .font(Tokens.Typography.bodyRelative)
                        .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                    if let flag = latest.flag {
                        FlagBadge(flag: flag)
                    }
                }
            }

            if let reference = trend.reference {
                Text("\(L10n.string("lab.trend.reference")): \(bandText(reference))")
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            } else if trend.points.contains(where: { $0.refLow != nil || $0.refHigh != nil }) {
                // Said outright rather than silently omitting the band: a
                // missing band otherwise reads as "no reference range".
                Text(L10n.string("lab.trend.rangesDiffer"))
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(Tokens.Palette.warning.resolve(for: scheme))
            }
        }
        .accessibilityElement(children: .combine)
    }

    private func bandText(_ band: ReferenceBand) -> String {
        switch (band.low, band.high) {
        case let (low?, high?): return "\(format(low)) – \(format(high)) \(trend.unit)"
        case let (nil, high?): return "< \(format(high)) \(trend.unit)"
        case let (low?, nil): return "> \(format(low)) \(trend.unit)"
        default: return "—"
        }
    }

    private func format(_ value: Double) -> String {
        value == value.rounded() ? String(Int(value)) : String(format: "%.2f", value)
    }
}

/// Critical values, above the charts and never inside one.
struct CriticalStrip: View {
    @Environment(\.colorScheme) private var scheme

    let results: [LabResult]

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.sm) {
            Label(L10n.string("lab.trend.criticalTitle"), systemImage: Tokens.State.labCritical.iconName)
                .font(Tokens.Typography.subheadingRelative)
                .foregroundStyle(Tokens.Palette.critical.resolve(for: scheme))

            ForEach(results) { result in
                Text("\(result.analyteName): \(result.value) \(result.unit)")
                    .font(Tokens.Typography.bodyRelative)
                    .foregroundStyle(Tokens.Palette.critical.resolve(for: scheme))
            }
        }
        .padding(Tokens.Spacing.md)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Tokens.Palette.criticalSurface.resolve(for: scheme))
        .clipShape(RoundedRectangle(cornerRadius: Tokens.Radius.md))
        .accessibilityElement(children: .combine)
    }
}

File: android/feature/lab-ui/src/main/kotlin/xyz/klinik/feature/lab/ui/LabTrendScreen.kt
package xyz.klinik.feature.lab.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import xyz.klinik.charts.ChartGeometry
import xyz.klinik.charts.Plot
import xyz.klinik.design.Tokens
import xyz.klinik.design.klinikColor
import xyz.klinik.feature.lab.LabTrendPhase
import xyz.klinik.feature.lab.LabTrendState
import xyz.klinik.network.AnalyteTrend
import xyz.klinik.network.LabFlag
import xyz.klinik.network.LabResult

/** Text the screen needs, resolved by the caller from string resources. */
data class LabTrendStrings(
    val empty: String,
    val notFound: String,
    val retry: String,
    val latest: String,
    val reference: String,
    val rangesDiffer: String,
    val criticalTitle: String,
    val flagName: (LabFlag) -> String,
    val message: (String) -> String,
)

@Composable
fun LabTrendScreen(
    state: LabTrendState,
    strings: LabTrendStrings,
    onRetry: () -> Unit,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        when (val phase = state.phase) {
            LabTrendPhase.Loading -> Centered { CircularProgressIndicator() }

            LabTrendPhase.Empty -> Centered {
                Text(strings.empty, color = klinikColor("textSecondary"))
            }

            LabTrendPhase.NotFound -> Centered {
                Text(strings.notFound, color = klinikColor("textSecondary"))
            }

            is LabTrendPhase.Failed -> Centered {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
                ) {
                    Text(strings.message(phase.messageKey), color = klinikColor("critical"))
                    TextButton(onClick = onRetry) { Text(strings.retry) }
                }
            }

            LabTrendPhase.Loaded -> Loaded(state, strings, onSelect)
        }
    }
}

@Composable
private fun Loaded(
    state: LabTrendState,
    strings: LabTrendStrings,
    onSelect: (String) -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(Tokens.Spacing.lg),
        verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
    ) {
        // Above the charts, always. A critical value a doctor has to go looking
        // for is one they can miss.
        if (state.critical.isNotEmpty()) {
            CriticalStrip(state.critical, strings)
        }

        AnalytePicker(state, onSelect)

        state.selectedTrend?.let { trend ->
            TrendChart(trend)
            TrendSummary(trend, strings)
        }
    }
}

@Composable
private fun AnalytePicker(state: LabTrendState, onSelect: (String) -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs)) {
        state.trends.forEach { trend ->
            val label = "${trend.analyteName} (${trend.unit})"

            if (trend.id == state.selectedTrend?.id) {
                Button(
                    onClick = { onSelect(trend.id) },
                    modifier = Modifier.fillMaxWidth().height(Tokens.minimumTouchTarget),
                ) {
                    Text(label)
                }
            } else {
                TextButton(
                    onClick = { onSelect(trend.id) },
                    modifier = Modifier.fillMaxWidth().height(Tokens.minimumTouchTarget),
                ) {
                    Text(label)
                }
            }
        }
    }
}

@Composable
private fun TrendChart(trend: AnalyteTrend) {
    val line = klinikColor("accent")
    val band = klinikColor("successSurface")
    val border = klinikColor("border")
    val critical = klinikColor("critical")
    val warning = klinikColor("warning")

    val reference = trend.reference?.let { ref ->
        val low = ref.low
        val high = ref.high
        if (low != null && high != null) low..high else null
    }

    val plot = ChartGeometry.plot(trend.points.map { it.value }, reference = reference)
    val flags = trend.points.map { it.flag }

    Canvas(
        modifier = Modifier
            .fillMaxWidth()
            .height(220.dp)
            // A plot is invisible to a screen reader however it is drawn, so
            // the summary below carries the same information in words and the
            // canvas itself is skipped.
            .clearAndSetSemantics {},
    ) {
        drawTrend(plot, flags, line, band, border, critical, warning)
    }
}

/**
 * The series, with its reference band behind it.
 *
 * The band is drawn only when the backend said every point shares one range.
 * Drawing the latest range across older points would put them on the wrong
 * side of a line they were never compared to.
 */
private fun DrawScope.drawTrend(
    plot: Plot,
    flags: List<LabFlag?>,
    line: Color,
    band: Color,
    border: Color,
    critical: Color,
    warning: Color,
) {
    if (plot.points.isEmpty()) return

    plot.band?.let {
        drawRect(
            color = band,
            topLeft = Offset(0f, (it.topY * size.height).toFloat()),
            size = Size(size.width, ((it.bottomY - it.topY) * size.height).toFloat()),
        )
    }

    fun at(index: Int) = Offset(
        (plot.points[index].x * size.width).toFloat(),
        (plot.points[index].y * size.height).toFloat(),
    )

    plot.points.indices.forEach { index ->
        if (index > 0) {
            drawLine(
                color = line,
                start = at(index - 1),
                end = at(index),
                strokeWidth = 2.dp.toPx(),
            )
        }

        // Colour marks the out-of-range points; the summary underneath says so
        // in words, because colour alone carries nothing for a reader who
        // cannot distinguish it (spec section 7).
        val colour = when (flags.getOrNull(index)) {
            LabFlag.CRITICAL -> critical
            LabFlag.LOW, LabFlag.HIGH -> warning
            else -> line
        }

        drawCircle(color = colour, radius = 4.dp.toPx(), center = at(index))
    }

    drawRect(color = border.copy(alpha = 0.4f), style = Stroke(width = 1.dp.toPx()))
}

@Composable
private fun TrendSummary(trend: AnalyteTrend, strings: LabTrendStrings) {
    val latest = trend.points.lastOrNull()

    Column(verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs)) {
        latest?.let { point ->
            val flag = point.flag?.let { ", ${strings.flagName(it)}" } ?: ""

            Text(
                "${strings.latest}: ${point.value} ${trend.unit}$flag",
                color = klinikColor("textPrimary"),
                modifier = Modifier.semantics {
                    contentDescription = "${trend.analyteName}, ${strings.latest} ${point.value} ${trend.unit}$flag"
                },
            )
        }

        val reference = trend.reference

        if (reference?.low != null && reference.high != null) {
            Text(
                "${strings.reference}: ${reference.low} – ${reference.high} ${trend.unit}",
                color = klinikColor("textSecondary"),
            )
        } else if (trend.points.any { it.refLow != null || it.refHigh != null }) {
            // Said outright rather than silently omitting the band: a missing
            // band otherwise reads as "no reference range".
            Text(strings.rangesDiffer, color = klinikColor("warning"))
        }
    }
}

/** Critical values, above the charts and never only inside one. */
@Composable
private fun CriticalStrip(results: List<LabResult>, strings: LabTrendStrings) {
    Surface(
        color = klinikColor("criticalSurface"),
        contentColor = klinikColor("critical"),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Column(
            modifier = Modifier.padding(Tokens.Spacing.md),
            verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs),
        ) {
            Text(strings.criticalTitle)

            results.forEach { result ->
                Row(horizontalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm)) {
                    Text("${result.analyteName}: ${result.value} ${result.unit}")
                }
            }
        }
    }
}

@Composable
private fun Centered(content: @Composable () -> Unit) {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { content() }
}
